src/proto/extension.rs

```diff
@@ -1,14 +1,81 @@
-use ssh_encoding::{Decode, Reader};
+//! SSH agent protocol extensions (messages & key constraints)
+
+use ssh_encoding::{CheckedSum, Decode, Encode, Error as EncodingError, Reader, Writer};
 use ssh_key::{public::KeyData, Signature};
 
-/// session-bind@openssh.com extension
+use super::ProtoError;
+
+// Reserved fields are marked with an empty string
+const RESERVED_FIELD: &str = "";
+
+/// SSH agent protocol *message* extension
+///
+/// Described in [draft-miller-ssh-agent-14 § 3.8](https://www.ietf.org/archive/id/draft-miller-ssh-agent-14.html#section-3.8)
+pub trait MessageExtension: 'static + Encode + Decode {
```

Comment:
I'm not sure about the enforcement of `Encode` and `Decode` here.

The `details` field in `Unparsed` but user might be free to encode that however they want.
They don't have to encode that according to SSH scheme. (Disclaimer: I use [postcard](https://crates.io/crates/postcard))

Reply:
Good pickup; I've moved these trait requirements to the `Extension::new_*{_raw}` methods instead

Reply:
That would allow, say:
```rust
pub struct ExtensionMessage(Vec<u8>);
impl Into<Vec<u8>> for ExtensionMessage { }
impl MessageExtension for QueryResponse {
    const NAME: &'static str = "message@example.com";
}
```

And then calling:
```
let message = ExtensionMessage(raw);
let extension = Extension::new_message_raw(message);
```

Not set on the API, but that way you can just pass in anything that has `Into<Vec<u8>>` implemented. Though, with `Unparsed` having a public vector inside of it, you can skip that if you want.